### PR TITLE
Trace payload chunking

### PR DESCRIPTION
**Pull request by marcotc**
During serialization, we now break down large collections of traces into smaller batches.

This is necessary because sending large payloads can cause the receiving server to reject them.
Currently, we send traces to the Datadog agent which has a [limit of 10 MiB (as of v6.14.1) per payload](https://github.com/DataDog/datadog-agent/blob/6.14.1/pkg/trace/api/api.go#L46).

We then break down traces into chunks that are smaller than that limit.
We also discard single traces that exceed that limit, as these cannot be broken down any further.

### CODE REVIEW

File: lib/ddtrace/transport/http/client.rb [hunk 1]
```diff
@@ -23,19 +23,19 @@ def send_request(request, &block)
           # Build request into env
           env = build_env(request)
 
-          # Get response from API
-          response = yield(current_api, env)
+          # Get responses from API
+          responses = yield(current_api, env)
```
Comment (delner) [inline lib/ddtrace/transport/http/client.rb]: Is there a way to call the client multiple times for each request in the batch instead of changing the client to do batching itself? I think it's unwise to make the client handle multiple requests simultaneously as it will greatly increase the complexity of the transport code.

It will also make this batching feature more brittle and tightly coupled to how HTTP works instead of being agnostic to the means of transport, which will make it difficult (if not impossible) to adopt new means of transport in the future.

File: lib/ddtrace/encoding.rb [hunk 2]
```diff
@@ -10,19 +10,84 @@ def content_type
         raise NotImplementedError
       end
 
-      # Encodes a list of traces, expecting a list of items where each items
-      # is a list of spans. Before dump the string in a serialized format all
-      # traces are normalized. The traces nesting is not changed.
-      def encode_traces(traces)
-        to_send = []
+      # Trace agent limit payload size of 10 MiB (as of agent v6.14.1):
```
[the remark this answers is not in the document]
Comment (marcotc) [reply lib/ddtrace/encoding.rb]: Found it. Since `5.11.0`: https://github.com/DataDog/datadog-trace-agent/blob/5.11.0/agent/receiver.go#L26

File: lib/ddtrace/transport/http/client.rb (hunk 1 above)
Comment (marcotc) [reply lib/ddtrace/transport/http/client.rb]: We need to know what encoder we are using in order to break down the `traces` being flushed into multiple chunks.

The client is currently responsible for such information, in the form of `client.current_api.spec.traces.encoder`.
Also, when downgrading, the encoder might change. Downgrading is currently handled by the client.

I tried to prototype a different aproach just now, moving the chunking logic as far up the call chain as I believe it makes sense: https://github.com/DataDog/dd-trace-rb/compare/feat/subdivide-payloads...tmp-feat/subdivide-payloads

I still don't like this one, too many layers are mixed together.

The main issue is that chunking, in a perfect scenario, would be done before we start calling the client. But the fact that we need the encoder, which is 2 levels down (inside the current `Spec` instance) and that the encoder can change if we need to downgrade the API, seem to make it quite tricky.

Next, I'm going to try to move `current_api` into the `transport` instance, and handle API versioning there, including the downgrading logic.

I'll report back on those results.

Comment (delner) [reply lib/ddtrace/transport/http/client.rb]: Okay, that context is helpful, thanks for that explanation.

The existing design had certain assumptions about encoding before, hence why it was buried down lower in the transport, as it was considered a detail of the current API, which I think still holds true to a great extent.

I think it brings up some legitimate questions about how the design could change to accommodate batching though. Some possible paradigms I can think of might be:

1. Expose the encoder, wrap the client with some kind of `Batcher`, then have the batcher encapsulate this logic entirely, and use the client to drive individual requests. `Batching` could be its own module `Batching` that can be composed into the existing `HTTP::Client`.
2. Assert that encoding requests is detail of the API and that its acceptable for the API to split requests on the client's behalf. Consequently, you'd make the API spec responsible for batching and splitting one large request into smaller ones (which is what I think you were effectively doing.)

There might be more ways of handling this, but the key differences between these basically is option 1 is explicit in `Client` usage (one request, one response) and option 2 is auto-magic "don't worry about the details, we'll figure it out."

Personally I'm in favor of number 1, because it keeps the responsibilities of the API/Client as small as possible (less complexity), and doesn't get us into weird scenarios where we have to handle a request that was forked into multiple requests in code that isn't concerned with batching (e.g. `Client#send_request`). Instead, we can keep all this batching code (hopefully) in a neat little module that knows how to deal with multiple requests and extends the capability of the `Client` in a compartmentalized way. (We could even go a step further to extract the "retry" functionality into a similar module for consistency, something I might want to undertake anyways.)

Let me know your thoughts or if you have some alternative paradigms to suggest!

File: lib/ddtrace/transport/http/transport.rb [hunk 3]
```diff
@@ -0,0 +1,92 @@
+module Datadog
```
Comment (marcotc) [inline lib/ddtrace/transport/http/transport.rb]: Ready for review again; the introduction of this class is the main change.
It adds a layers between the `writer` and `client` that handles chunking and API downgrading.

**Comment by marcotc** [comment]
Successfully tested locally with a few real example applications.

**Comment by marcotc** [comment]
Updated with master, ready for review. (Don't mind Rails 5.2.1.4 breaking the Ruby 2.2 build 😞)

File: lib/ddtrace/transport/traces.rb [hunk 4]
```diff
@@ -8,25 +9,173 @@ module Traces
...
+          encoder.encode(trace.map(&:to_hash))
+        end
+      end
+
+      # Sends traces based on transport API configuration.
+      #
+      # This class initializes the HTTP client, breaks down large
+      # batches of traces into smaller chunks and handles
+      # API version downgrade handshake.
+      class Transport
```
Comment (marcotc) [inline lib/ddtrace/transport/traces.rb:101]: I move the transport to this `traces.rb` file, as it currently is tracer-specific. This should allow for a more clear separation from other tracer-agnostic parts.

File: lib/ddtrace/transport/traces.rb [hunk 5]
```diff
@@ -8,25 +9,173 @@ module Traces
...
+        # Encodes a list of traces in chunks.
+        # Before serializing, all traces are normalized. Trace nesting is not changed.
+        #
+        # @param traces [Array<Trace>] list of traces
+        # @return [Enumerator[Array[Bytes,Integer]]] list of encoded chunks: each containing a byte array and
+        #   number of traces
+        def encode_in_chunks(traces)
+          encoded_traces = traces.map { |t| encode_one(t) }.compact
+
+          Datadog::Chunker.chunk_by_size(encoded_traces, max_size).map do |chunk|
```
Comment (delner) [inline lib/ddtrace/transport/traces.rb:65]: I like the simplicity here; very nice!

File: lib/ddtrace/transport/traces.rb [hunk 6]
```diff
@@ -8,25 +9,173 @@ module Traces
...
+        end
+
+        # Encodes a list of traces in chunks.
+        # Before serializing, all traces are normalized. Trace nesting is not changed.
+        #
+        # @param traces [Array<Trace>] list of traces
+        # @return [Enumerator[Array[Bytes,Integer]]] list of encoded chunks: each containing a byte array and
+        #   number of traces
+        def encode_in_chunks(traces)
+          encoded_traces = traces.map { |t| encode_one(t) }.compact
```
Comment (delner) [inline lib/ddtrace/transport/traces.rb:63]: Not something that must be resolved in this PR but something to think about: should encoding be done as chunks are generated and submitted?

My thinking is if we spend the CPU time on encoding a batch of traces, we should definitely attempt sending it, as we've already paid the cost for it. I think there's a few of ways to go about this:

 - On each chunk, do the encoding only for that chunk and `yield` back to do a request (similar to what you had before.)
 - Move the use of `Chunker` entirely outside the transport; on each chunk, `yield` back a chunk of traces, then `send_traces` with that chunk to the transport. This might be a bigger change because on each chunk, you'd presumably have to encode first before `send_traces`, so we'd have to change encoding based on what encoder is active on the transport at the time. (That would involve extracting a lot of encoding behavior out of the transport.)

Ultimately I think it'd make more sense to end up somewhere like the second suggestion, but the first might be an okay stopgap. Neither though have to be done in this PR... just a consideration for the future.

File: lib/ddtrace/transport/traces.rb (hunk 4 above)
Comment (delner) [reply lib/ddtrace/transport/traces.rb:101]: I think this is okay for now, but conceivably, other data types might want to re-use API behavior in the transport somewhere down the road. Still I like the simplification of the transport itself, and how we're separating concerns, so I'm happy with this change right now.

File: lib/ddtrace/transport/traces.rb [hunk 7]
```diff
@@ -8,25 +9,173 @@ module Traces
...
+          @apis = apis
+          @default_api = default_api
+
+          change_api!(default_api)
+        end
+
+        def send_traces(traces)
+          encoder = current_api.encoder
+
+          chuncker = Datadog::Transport::Traces::Chunker.new(encoder)
```
Comment (delner) [inline lib/ddtrace/transport/traces.rb:114]: `chuncker` is a mispelling.

File: spec/ddtrace/chunker_spec.rb [hunk 8]
```diff
@@ -0,0 +1,24 @@
...
+    let(:max_chunk_size) { 3 }
+
+    it do
+      expect(subject.to_a).to eq([%w[1 22], ['333']])
+    end
+
+    context 'with single element that is too large' do
+      let(:list) { ['55555'] }
+
+      it 'returns single element exceeding maximum' do
```
Comment (delner) [inline spec/ddtrace/chunker_spec.rb:19]: Not sure I understand why this behavior is, or how we work with it?

- Are we avoiding this scenario entirely by using `encode_one` and `compact`? Rendering this moot for the trace chunking?
- Not to say we shouldn't, but is there a particular reason we should return this single element? Are there other choices we could consider as well?

File: lib/ddtrace/chunker.rb [hunk 9]
```diff
@@ -0,0 +1,37 @@
...
+    # element will be returned in its own chunk. You have to verify by yourself
+    # when such elements are returned.
+    #
+    # @param list [Enumerable] list of elements
+    # @param max_chunk_size [Numeric] maximum acceptable chunk size
+    # @return [Enumerator] lazy list of chunks
+    def chunk_by_size(list, max_chunk_size)
+      chunk_agg = 0
+      list.slice_before do |elem|
+        size = elem.size
```
Comment (delner) [inline lib/ddtrace/chunker.rb:21]: There's an assumption that every item in the list responds to `size`; is this too specific?

File: lib/ddtrace/transport/traces.rb (hunk 6 above)
Comment (brettlangdon) [reply lib/ddtrace/transport/traces.rb:63]: The important piece is that we need to encode each trace individually to know how it'll fit into the allowed max buffer size, and then join them together after based on the encoding method used.

I think the way this reads to me is:

- Encode each trace individually
- Group the traces into chunks based on the max allowed payload size
- Send all chunks to the agent
  - Regardless of how many traces are in the chunks, e.g. sending 1 trace yields 1 chunk and 1 request to the agent

Which if ☝️ is right then it is correct behavior.

We could probably optimize for, encode and generate chunks, and send a request before encoding any more traces to do minimal encoding overhead between requests, but it shouldn't matter too much. Just whether we encode all up front then send, or encode in chunks as we make requests, should be roughly the same CPU overhead either way.

File: lib/ddtrace/transport/traces.rb [hunk 10]
```diff
@@ -8,25 +9,173 @@ module Traces
...
+          chuncker.encode_in_chunks(traces).map do |encoded_traces, trace_count|
+            request = Datadog::Transport::Traces::Request.new(encoded_traces, trace_count)
+
+            client.send_payload(request).tap do |response|
+              if downgrade?(response)
+                downgrade!
+                return send_traces(traces)
```
Comment (brettlangdon) [inline lib/ddtrace/transport/traces.rb:121]: Ruby question: Does this `return` return on the `send_traces` method or on the map or tap blocks?

File: lib/ddtrace/writer.rb [hunk 11]
```diff
@@ -72,20 +72,19 @@ def send_spans(traces, transport)
...
-        unless priority_sampler.nil? || response.service_rates.nil?
-          priority_sampler.update(response.service_rates)
-        end
+      # Tally up successful flushes
+      responses.reject { |x| x.internal_error? || x.server_error? }.each do |response|
+        @traces_flushed += response.trace_count
       end
 
+      # Update priority sampler
+      update_priority_sampler(responses.last)
```
Comment (brettlangdon) [inline lib/ddtrace/writer.rb:84]: What if there are no successful responses?

Comment (brettlangdon) [reply lib/ddtrace/writer.rb:84]: I see `update_priority_sampling_sampler` might be handling this?

File: lib/ddtrace/transport/traces.rb (hunk 6 above)
Comment (marcotc) [reply lib/ddtrace/transport/traces.rb:63]: I can easily change it to a stream model:
1. Encode as many traces as it fits in a chunk
2. Send chunk
3. Repeat 1 with next chunk

I didn't add it to the PR changes for simplicity (as we already kept all traces serialized in memory before, so the PR is not increasing overahead), but I can add streaming now.

File: spec/ddtrace/chunker_spec.rb (hunk 8 above)
Comment (marcotc) [reply spec/ddtrace/chunker_spec.rb:19]: It's a matter of what happens when you ask for chunks of max_size 10, but you give it a single element of size 11?

I'm returning the oversized element here because then the caller can make an informed decision (and possibly output a message or record a metric) about this oversized element. There's no extra cost in returning the existing element, instead of discarding it, so I chose to return it and let the caller handle it.

This test asserts that this edge-case behaves the way we want.

File: lib/ddtrace/chunker.rb (hunk 9 above)
Comment (marcotc) [reply lib/ddtrace/chunker.rb:21]: I tried to extract this logic and turned out that `chunk_by_size` retained no logic, just delegating methods with worse ergonomics than the caller simply calling `#slice_before` with the same logic as here.
I tried that, to move `#slice_before` to the caller, but then it became a little bit too complex at the caller site, so I chose to keep this specific implementation here, in order to encapsulate this logic.

When we have further use cases I think we'll be able to know how to extract it, as making it truly generic today didn't work quite as well.

File: lib/ddtrace/transport/traces.rb (hunk 10 above)
Comment (marcotc) [reply lib/ddtrace/transport/traces.rb:121]: It returns the nearest method definition, `def send_traces` in this case. `next` would return from this `tap` proc instead, for reference.

File: lib/ddtrace/writer.rb (hunk 11 above)
Comment (marcotc) [reply lib/ddtrace/writer.rb:84]: You are correct, this just a matter of extracting a somewhat complex logic block into its own method.

**Comment by marcotc** [comment]
One thing I noticed doing another pass at this PR is that many of the files under `lib/transport/transport` that have generic intent (e.g. `Datadog::Transport::HTTP::Client`) are actually trace-specific.
I move a few components that were trace-specific and directly affected by this PR into trace-specific namepaces. But there's definitely more work in order to make the Transport package have a clean separation between an abstract transport vs use-case specific logic.

File: lib/ddtrace/transport/traces.rb (hunk 6 above)
Comment (marcotc) [reply lib/ddtrace/transport/traces.rb]: Done by using a lazy iterator, which will process one request at a time, in a stream fashion, without ever creating intermediate lists.

The end result is still a list with all Response objects.

File: spec/ddtrace/chunker_spec.rb (hunk 8 above)
Comment (delner) [reply spec/ddtrace/chunker_spec.rb:19]: I see. It's a little strange, but if your only choices are return batches that are too big, or omit them entirely, I think this makes more sense.

File: lib/ddtrace/chunker.rb (hunk 9 above)
Comment (delner) [reply lib/ddtrace/chunker.rb:21]: Okay, I think that's fair. We can always mature this a bit as we see new cases appear.

**Comment by delner** [comment]
@marcotc Agreed; I'd like to see the trace-specific logic entirely extracted from the `Transport` namespace if at all possible, at some point in the future. More of this behavior probably belongs in `TraceWriter` instead.

#1014 did this for profiling; although it too needs to mature a bit more, I think how this example keeps the domain (profiling) logic together is the direction to go in.

However we don't need to undertake a major reorganization in this PR unless its a blocker to the payload chunking here. Use your judgment for what you think is reasonable.

**Comment by marcotc** [comment]
There were non-trivial changes to `Datadog::Transport::IO` that I did not expect to undertake in order to adapt it to use a few of the interfaces that have changed.

But otherwise, there is no part of this PR that was touched only for the sake of refactoring: all components touched required changes for correctness.

File: lib/ddtrace/chunker.rb [hunk 12]
```diff
@@ -22,7 +22,7 @@ def chunk_by_size(list, max_chunk_size)
         chunk_agg += size
         if chunk_agg > max_chunk_size
           # Can't fit element in current chunk
-          # TODO Datadog::Debug::HealthMetrics.increment('tracer.chunker.chunked')
+          Datadog.health_metrics.transport_chunked(1, tags: ["max_size:#{max_chunk_size}"])
```
Comment (delner) [inline lib/ddtrace/chunker.rb:25]: Not sure we need to emit a metric each time a chunk occurs. Is it possible to do it outside this function, say perhaps where `chunk_by_size` or `encode_in_chunks` is called, and emit one metric with the number of chunks in aggregate? Would give the insight necessary into how many chunks are produced without having a 1:1 number of metrics.

Comment (marcotc) [reply lib/ddtrace/chunker.rb]: Moved to bulk reporting now 👍

**Comment by gbbr** [comment]
I didn't read all of this, but note that if you split a trace in two it will break stats. I suspect the "payload too big" thing happens anyway with big traces so it would be necessary to do that. An implementation like that will complicate tracer code a lot I suspect, and it happens not only in Ruby but other languages too.

Instead, I think it might be a better to idea to explore "span batching" in the agent again, to add an endpoint which receives a set of random spans, and reconstructs traces on the agent side. This brings many benefits:

* Solves the payload size problem for all languages
* Significantly simplifies span buffering / trace batching in all clients, reducing memory and CPU fingerprint in the host application
* Allow supporting open standard like OpenTelemetry, Zipkin, etc.

This will of course move the memory problem into the trace-agent, from not one client, but multiple ones which may be sending to the same endpoint. It should be acceptable, but it's bound to bring new problems and complications and needs to be explored (again).

**Comment by brettlangdon** [comment]
@gbbr this particular change is to break payloads into smaller sizes by separating individual traces. It is not for breaking traces into smaller pieces.

e.g. if one flush interval passes and we have 10mb of traces, we'll send 2 payloads of traces to the agent instead of trying to do it in one.

This is what we do in a few of the languages now.

Span streaming is a great idea, but will require significant development/coordination between the tracers and the agent. This change should unblock us for now while we schedule span streaming investigation/work.

**Comment by gbbr** [comment]
Alright, carry on :) Never mind me then.

File: lib/ddtrace/transport/traces.rb [hunk 13]
```diff
@@ -8,25 +9,177 @@ module Traces
...
+          encoder = current_api.encoder
+          chunker = Datadog::Transport::Traces::Chunker.new(encoder)
+
+          responses = chunker.encode_in_chunks(traces.lazy).map do |encoded_traces, trace_count|
+            request = Datadog::Transport::Traces::Request.new(encoded_traces, trace_count)
+
+            client.send_payload(request).tap do |response|
+              if downgrade?(response)
+                downgrade!
+                return send_traces(traces.lazy)
```
Comment (delner) [inline lib/ddtrace/transport/traces.rb:121]: What's `traces.lazy`?

When we fail, downgrade and retry, are we re-encoding already encoded output, or doing chunking again? Unless I'm misunderstanding something, I would expect this probably to be `send_payload` no?

File: lib/ddtrace/transport/traces.rb [hunk 14]
```diff
@@ -8,25 +9,177 @@ module Traces
...
+              end
+            end
+          end.force
```
Comment (delner) [inline lib/ddtrace/transport/traces.rb:124]: What's `force`?

File: lib/ddtrace/transport/traces.rb [hunk 15]
```diff
@@ -8,25 +9,177 @@ module Traces
...
+
+          Datadog.health_metrics.transport_chunked(responses.size, tags: ["max_size:#{chunker.max_size}"])
```
Comment (delner) [inline lib/ddtrace/transport/traces.rb:126]: What's the difference between `size` and `max_size` and what does it tell us? Just want to know what benefit this extra tag provides.

File: lib/ddtrace/transport/traces.rb (hunk 13 above)
Comment (marcotc) [reply lib/ddtrace/transport/traces.rb]: This one is unnecessary, as the recursive call will already do `#lazy` a few lines above, nice catch!

File: lib/ddtrace/transport/traces.rb (hunk 15 above)
Comment (marcotc) [reply lib/ddtrace/transport/traces.rb:126]: In case the user wants to tweak the `max_size`, they can track how that affects the numbers of chunks generated in their application.
It's a long-shot, tbh, I'm not strongly attached to this tag.

Comment (delner) [reply lib/ddtrace/transport/traces.rb:126]: If the user tweaks the max size and we tag it, do you have a use case in mind for how we'd use this information?

I think if we can say "it'll help us determine whether chunks are too big when nothing else will", or there's some other articulated case for how we use the tag, I think the tag is probably fine to add. Otherwise probably better to omit. I'm on the fence.

Comment (brettlangdon) [reply lib/ddtrace/transport/traces.rb:126]: please do not tag with max size. This is not a good tag. The cardinality on it is unbounded, and its usability in dashboards/monitors won't be any good.

Instead consider creating separate metrics for number of responses, and the number of requests sent/chunks created/etc.

File: lib/ddtrace/transport/traces.rb [hunk 16]
```diff
@@ -8,25 +9,177 @@ module Traces
...
+
+        private
+
+        def encode_one(trace)
+          encoded = Encoder.encode_trace(encoder, trace)
+
+          if encoded.size > max_size
+            # This single trace is too large, we can't flush it
+            Datadog.logger.debug { "Dropping trace. Payload too large: '#{trace.map(&:to_hash)}'" }
+            Datadog.health_metrics.transport_trace_too_large(1, tags: ["max_size:#{max_size}"])
```
Comment (brettlangdon) [inline lib/ddtrace/transport/traces.rb:78]: do not tag with `max_size`

File: lib/ddtrace/transport/traces.rb [hunk 17]
```diff
@@ -8,25 +9,177 @@ module Traces
       class Parcel
         include Transport::Parcel
 
-        def count
-          data.length
+        attr_reader :trace_count
+
+        def initialize(data, trace_count)
```
Comment (delner) [inline lib/ddtrace/transport/traces.rb:14]: I don't think this makes sense with the current design, and altering this has a lot of side effects.

`Traces::Parcel` in the current design is supposed to be a protocol agnostic package of trace data, to be created by something that doesn't have knowledge of how the transport works. By requiring the `Parcel` to be given encoded data like this (along with its trace count), it implicitly requires a knowledge of the transport and its current API state to properly construct, rendering this an object with strong coupling to internal transport behavior.

That said, I think there's an argument to be made that we should change the design, and that to support encoding in chunks, it might also require some different kind of construct.

Short term, maybe we can remedy this by leaving the existing `Traces::Parcel` as is, but creating a new `Traces::EncodedParcel` which results from encoding traces from a parcel during chunking.

Long term, I think perhaps it shouldn't create parcels at all; the transport should only receive generic requests, and return generic responses... any trace specific behavior should be in some kind of `Traces::Transport` that wraps a generic transport (HTTP/IO/UDS, etc.) Then we shouldn't have a need for any parcels.

File: lib/ddtrace/transport/traces.rb (hunk 16 above)
Comment (marcotc) [reply lib/ddtrace/transport/traces.rb]: 👌, removed.

File: lib/ddtrace/transport/traces.rb (hunk 15 above)
Comment (marcotc) [reply lib/ddtrace/transport/traces.rb]: It should be always the same value today for every client, and it's fixed once the application starts, but I've removed it as per recommendation.

File: lib/ddtrace/transport/traces.rb (hunk 17 above)
Comment (marcotc) [reply lib/ddtrace/transport/traces.rb:14]: The reason I changed this Parcel is because this Parcel is actually a subtype, a `Traces::Parcel`, it inherits behaviour from the generic parcel `Transport::Parcel`, which is agnostic.

I did have the additional information I needed (`trace_count`) under another carrier object, which I believe was `Traces::Request`, but the parcel seemed like a better carrier for it. To be fair, I'm not 100% sure on the role of Parcel after the changes, so I'm very much open to changing this design.
